File: Python/5-BinarySearch/Problems/1-BinarySearch/BinarySearch.py
# Binary Search: 5 Different Approaches
# Binary search is an efficient algorithm for finding elements in a sorted array.

from typing import List # Import List for type hinting, making the code more readable
import bisect # Import the bisect module, which provides efficient implementations of binary search

# 1. Standard Iterative Binary Search
def binary_search_iterative(arr: List[int], target: int) -> int:
    """
    Performs iterative binary search on a sorted list to find the index of a target value.

    Args:
        arr: A sorted list of integers.
        target: The integer value to search for.

    Returns:
        The index of the target value in the list if found, otherwise -1.

    Example:
        arr = [1, 3, 5, 7, 9, 11, 13]
        target = 7
        binary_search_iterative(arr, target) == 3
    """
    left, right = 0, len(arr) - 1 # Initialize left and right pointers to the start and end of the list
    while left <= right: # Continue searching as long as the left pointer is less than or equal to the right pointer
        mid = left + (right - left) // 2 # Calculate the middle index.  This avoids potential overflow issues with large lists.
        if arr[mid] == target:
            return mid # If the middle element is the target, return its index
        elif arr[mid] < target:
            left = mid + 1 # If the middle element is less than the target, search the right half of the list
        else:
            right = mid - 1 # If the middle element is greater than the target, search the left half of the list
    return -1 # If the target is not found, return -1

# 2. Recursive Binary Search
def binary_search_recursive(arr: List[int], left: int, right: int, target: int) -> int:
    """
    Performs recursive binary search on a sorted list.

    Args:
        arr: A sorted list of integers.
        left: The starting index of the search interval.
        right: The ending index of the search interval.
        target: The integer value to search for.

    Returns:
        The index of the target value in the list if found, otherwise -1.
    """
    if left > right: # Base case: if the left pointer crosses the right pointer, the target is not found
        return -1
    mid = left + (right - left) // 2 # Calculate the middle index
    if arr[mid] == target:
        return mid # If the middle element is the target, return its index
    elif arr[mid] < target:
        return binary_search_recursive(arr, mid + 1, right, target) # Recursively search the right half
    else:
        return binary_search_recursive(arr, left, mid - 1, target) # Recursively search the left half

# 3. Binary Search Using Python's Bisect Module
def binary_search_bisect(arr: List[int], target: int) -> int:
    """
    Performs binary search using Python's built-in bisect_left function.
    The bisect module provides efficient implementations of binary search.

    Args:
        arr: A sorted list of integers.
        target: The integer value to search for.

    Returns:
        The index of the target value in the list if found, otherwise -1.
    """
    index = bisect.bisect_left(arr, target) # Use bisect_left to find the insertion point for the target
    if index < len(arr) and arr[index] == target: # Check if the element at the insertion point is the target
        return index
    return -1 # If the target is not found, return -1

# 4. Binary Search to Find the First Occurrence
def binary_search_first_occurrence(arr: List[int], target: int) -> int:
    """
    Finds the index of the first occurrence of a target value in a sorted list.

    Args:
        arr: A sorted list of integers.
        target: The integer value to search for.

    Returns:
        The index of the first occurrence of the target value, or -1 if not found.
    """
    left, right = 0, len(arr) - 1
    result = -1 # Initialize the result to -1 (not found)
    while left <= right:
        mid = left + (right - left) // 2
        if arr[mid] == target:
            result = mid # When found, update result, and continue searching the left side
            right = mid - 1  # Keep searching on the left side for the first occurrence
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return result

# 5. Binary Search to Find the Last Occurrence
def binary_search_last_occurrence(arr: List[int], target: int) -> int:
    """
    Finds the index of the last occurrence of a target value in a sorted list.

    Args:
        arr: A sorted list of integers.
        target: The integer value to search for.

    Returns:
        The index of the last occurrence of the target value, or -1 if not found.
    """
    left, right = 0, len(arr) - 1
    result = -1 # Initialize result to -1
    while left <= right:
        mid = left + (right - left) // 2
        if arr[mid] == target:
            result = mid # update the result
            left = mid + 1  # Keep searching on the right side for the last occurrence
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return result

# Sample Data
arr = [1, 3, 5, 5, 5, 7, 9, 11, 13]
target = 5

# Running Tests
print("Iterative Binary Search:", binary_search_iterative(arr, target))
print("Recursive Binary Search:", binary_search_recursive(arr, 0, len(arr) - 1, target))
print("Bisect Binary Search:", binary_search_bisect(arr, target))
print("First Occurrence Search:", binary_search_first_occurrence(arr, target))
print("Last Occurrence Search:", binary_search_last_occurrence(arr, target))

File: Java/5-BinarySearch/Problems/1-BinarySearch/BinarySearch.java
import java.util.Arrays;

public class BinarySearch {

    /**
     * **1. Iterative Binary Search**
     *
     * Performs a binary search on a sorted array using iteration.
     *
     * **Parameters:**
     * - `arr`: The sorted integer array to search.
     * - `target`: The integer value to search for.
     *
     * **Returns:**
     * The index of the `target` if found, otherwise -1.
     *
     * **Time Complexity:** O(log n) - logarithmic time complexity.
     * **Space Complexity:** O(1) - constant space complexity.
     */
    public static int iterativeBinarySearch(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2; // Prevent potential overflow
            // Check if the target is present at the middle
            if (arr[mid] == target) {
                return mid;
            }
            // If target is greater, ignore left half
            if (arr[mid] < target) {
                left = mid + 1;
            }
            // If target is smaller, ignore right half
            else {
                right = mid - 1;
            }
        }
        // If we reach here, the element was not found
        return -1;
    }

    /**
     * **2. Recursive Binary Search**
     *
     * Performs a binary search on a sorted array using recursion.
     *
     * **Parameters:**
     * - `arr`: The sorted integer array to search.
     * - `target`: The integer value to search for.
     * - `left`: The starting index of the search range.
     * - `right`: The ending index of the search range.
     *
     * **Returns:**
     * The index of the `target` if found, otherwise -1.
     *
     * **Time Complexity:** O(log n) - logarithmic time complexity.
     * **Space Complexity:** O(log n) - due to the recursive call stack.
     */
    public static int recursiveBinarySearch(int[] arr, int target, int left, int right) {
        if (left > right) {
            return -1; // Base case: element not found
        }

        int mid = left + (right - left) / 2;  // Prevent potential overflow

        if (arr[mid] == target) {
            return mid; // Base case: element found
        }

        if (arr[mid] < target) {
            return recursiveBinarySearch(arr, target, mid + 1, right); // Search right half
        } else {
            return recursiveBinarySearch(arr, target, left, mid - 1); // Search left half
        }
    }

    /**
     * **3. Binary Search using Java's `Arrays.binarySearch()`**
     *
     * Utilizes Java's built-in `Arrays.binarySearch()` method.
     *
     * **Parameters:**
     * - `arr`: The sorted integer array to search.
     * - `target`: The integer value to search for.
     *
     * **Returns:**
     * The index of the `target` if found, otherwise a negative value
     * (-(insertion point) - 1).
     *
     * **Time Complexity:** O(log n) - logarithmic time complexity.
     * **Space Complexity:** O(1) - constant space complexity.
     */
    public static int javaBinarySearch(int[] arr, int target) {
        return Arrays.binarySearch(arr, target);
    }

    /**
     * **4. Binary Search to find the First Occurrence**
     *
     * Finds the index of the *first* occurrence of the target value in a sorted array
     * which may contain duplicate elements.
     *
     * **Parameters:**
     * - `arr`: The sorted integer array to search.
     * - `target`: The integer value to search for.
     *
     * **Returns:**
     * The index of the *first* occurrence of the `target`, or -1 if not found.
     *
     * **Time Complexity:** O(log n)
     * **Space Complexity:** O(1)
     */
    public static int findFirstOccurrence(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;
        int result = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (arr[mid] == target) {
                result = mid;       // Found a match, but look for earlier ones
                right = mid - 1;    // Continue searching in the left half
            } else if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return result;
    }

    /**
     * **5. Binary Search to find the Last Occurrence**
     *
     * Finds the index of the *last* occurrence of the target value in a sorted array
     * which may contain duplicate elements.
     *
     * **Parameters:**
     * - `arr`: The sorted integer array to search.
     * - `target`: The integer value to search for.
     *
     * **Returns:**
     * The index of the *last* occurrence of the `target`, or -1 if not found.
     *
     * * **Time Complexity:** O(log n)
     * * **Space Complexity:** O(1)
     */
    public static int findLastOccurrence(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;
        int result = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (arr[mid] == target) {
                result = mid;       // Found a match, but look for later ones
                left = mid + 1;     // Continue searching in the right half
            } else if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[] sortedArray = {2, 5, 8, 12, 16, 23, 38, 56, 72, 91};
        int targetValue = 23;

        // 1. Iterative Binary Search
        int iterativeResult = iterativeBinarySearch(sortedArray, targetValue);
        System.out.println("Iterative Binary Search: Index of " + targetValue + " is " + iterativeResult);

        // 2. Recursive Binary Search
        int recursiveResult = recursiveBinarySearch(sortedArray, targetValue, 0, sortedArray.length - 1);
        System.out.println("Recursive Binary Search: Index of " + targetValue + " is " + recursiveResult);

        // 3. Java's Arrays.binarySearch()
        int javaSearchResult = javaBinarySearch(sortedArray, targetValue);
        System.out.println("Java's Arrays.binarySearch(): Index of " + targetValue + " is " + javaSearchResult);

        // Example with duplicates to test first and last occurrence
        int[] duplicateArray = {2, 4, 6, 6, 6, 8, 10, 12};
        int duplicateTarget = 6;

        // 4. Find First Occurrence
        int firstOccurrence = findFirstOccurrence(duplicateArray, duplicateTarget);
        System.out.println("First Occurrence of " + duplicateTarget + " is at index " + firstOccurrence);

        // 5. Find Last Occurrence
        int lastOccurrence = findLastOccurrence(duplicateArray, duplicateTarget);
        System.out.println("Last Occurrence of " + duplicateTarget + " is at index " + lastOccurrence);

        // Example when the target is not in the array
        int notFoundTarget = 15;
        int notFoundResult = iterativeBinarySearch(sortedArray, notFoundTarget);
        System.out.println("Index of " + notFoundTarget + " is " + notFoundResult); // Output: -1
    }
}

